### PR TITLE
Compiler: Add ability to test compiler

### PR DESCRIPTION
**Pull request by johnwparent**
Extends the compiler class to add a smoke test for a given compiler class + path to a compiler.
* Compiles either provided test files or Spack provided default templates
* Can test c/cxx/fortran
* Compilation is performed in a temp directory, all artifacts discarded
* Compilation language selected by caller
* Can work either with an instantiated compiler, or a class and a path to a compiler (i.e. for testing compilers when detecting or auditing)

Adds checks in the compilers audit, detection, and prebuild steps to verify compiler validity. Will stop a build if the compiler selected by the concertizer has broken components.

### CODE REVIEW

File: lib/spack/spack/compiler.py [hunk 1]
```diff
@@ -655,6 +667,65 @@ def search_regexps(cls, language):
...
+        test_file = file if file else default_try_file_for_lang(lang)
+        compiler = spack.util.executable.Executable(compiler_path)
+        compiler_env = cls._cmp_test_env(compiler_path)
+        compiler_invocation_args = {
+            "output": str,
+            "error": str,
+            "timeout": 120,
+            "fail_on_error": True,
+            "env": compiler_env,
+        }
```
Comment (scheibelp) [inline lib/spack/spack/compiler.py:709]: This should be running environment setup related to compiler (namely the same thing that `build_environment` does, like loading modules, setting env variables as defined in the compiler definition, and possibly other things).

File: share/spack/compiler/f77 [hunk 2]
```diff
@@ -0,0 +1 @@
+fc
```
Comment (scheibelp) [inline share/spack/compiler/f77:1]: (question) is this properly formatted? How is it used? It looks to me like `default_try_file_for_lang` would return the wrong thing for `"f77"`.

File: lib/spack/spack/build_environment.py [hunk 3]
```diff
@@ -317,6 +318,17 @@ def _add_werror_handling(keep_werror, env):
...
 
+def validate_compiler_for_pkg(pkg):
+    """Ensures selected compiler is actually capable of compiling
+    Fail early and provide an convenient warning in case it is not
+    """
+    compiler = pkg.compiler
+    for lang in ("cc", "cxx", "fc", "f77"):
+        comp = getattr(lang, compiler)
+        if comp:
+            compiler.try_compiler(lang)
```
Comment (scheibelp) [inline lib/spack/spack/build_environment.py:311]: I think this might have some conceptual overlap with `asp.c_compiler_runs -> compiler.Compiler.compiler_verbose_output`

Although `c_compiler_runs` generally is only run on `linux` - when `using_libc_compatibility() -> True`. I think it would be reasonable to "promote" that and have it always run.

File: lib/spack/spack/compiler.py [hunk 4]
```diff
@@ -655,6 +667,65 @@ def search_regexps(cls, language):
...
+        """
+
+        # if file is not set by caller, use default file for specified language
+        # if there is no specified language, we don't know what to compile, fail
+        determinable_file = file or lang
+        if not determinable_file:
+            err = "No file or language provided."
+            raise TryCompilerError(cls.__name__, file, lang, err)
+
+        test_file = file if file else default_try_file_for_lang(lang)
```
Comment (scheibelp) [inline lib/spack/spack/compiler.py:702]: We'll want to replicate the logic from `build_environment.setup_package` that ensures that all env vars that need to be defined for the compiler are set - that includes:

- `set_compiler_environment_variables`
-  `for mod in pkg.compiler.modules: load_module(mod)`

(and this should happen in a subprocess, or a context manager that undoes env var modifications).

File: lib/spack/spack/compiler.py [hunk 5]
```diff
@@ -711,6 +782,24 @@ def compiler_environment(self):
...
+    FC = "fc"
+    F77 = "f77"
+    ext_for_lang = {CC: "c", CXX: "cxx", FC: "f", F77: "f77"}
+
+
+class TryCompilerError(spack.error.SpackError):
+    def __init__(self, compiler_name, file=None, lang=None, err=None):
+        lang = lang if lang else "unknown"
+        file = file if file else "unknown"
+        error_stmt = f"Unable to compile even a simple test file ({file})\
```
Comment (scheibelp) [inline lib/spack/spack/compiler.py]: (minor) Can you reword this to actually just dump the file contents like

```
Unable to compile the following file:

...
```

File: lib/spack/spack/compilers/__init__.py [hunk 6]
```diff
@@ -902,10 +902,11 @@ def _default(fn_args):
                 return value, None
 
             error = f"Couldn't get version for compiler {path}".format(path)
-        except spack.util.executable.ProcessError as e:
-            error = f"Couldn't get version for compiler {path}\n" + str(e)
+        # ProcessTimeoutError is a subclass of ProcessError, needs to be caught first
```
Comment (scheibelp) [inline lib/spack/spack/compilers/__init__.py:905]: +1

**Review by scheibelp — CHANGES_REQUESTED** [review]

File: share/spack/compiler/f77 (hunk 2 above)
Comment (johnwparent) [reply share/spack/compiler/f77:1]: It was supposed to be a link to the file in `share/spack/compiler/fc` as I wasn't sure which Fotran standard to use for `fc` so I just used f77. I have 0 experience with Fortran, so could use some guidance as to whether this is the correct thing to do

File: lib/spack/spack/build_environment.py (hunk 3 above)
Comment (johnwparent) [reply lib/spack/spack/build_environment.py:311]: I agree, I think the logic in `asp.CompilerParser` could use some integration here and IMO should completely replace my changes from this PR in `build_environment.py` as this is what I was trying to achieve with those changes and this is objectively better.

Will combine `compiler.Compiler._compile_dummy_c_source` with `try_compiler` (`_compile_dummy_c_source` should probably merge with `try_compiler`) and will update the logic in `asp.CompilerParser` accordingly

File: lib/spack/spack/compiler.py (hunk 4 above)
Comment (johnwparent) [reply lib/spack/spack/compiler.py:702]: Good point.
I think `Compiler.compiler_environment` will handle the environment context as well as a good bit of the setup (it calls `load_module` under the hood), I'll just need to make sure we cover the rest of the setup.

**Review by alalazo — CHANGES_REQUESTED** [review]
Let's sync on this with @becker33 

`Compiler` objects are going away as soon as possible, and compilers will be "just" regular packages. PRs adding new code to `spack.compilers.*`, instead of the corresponding package, are going in the opposite direction.

File: lib/spack/spack/solver/asp.py [hunk 7]
```diff
@@ -2987,11 +2994,7 @@ class CompilerParser:
     def __init__(self, configuration) -> None:
         self.compilers: Set[KnownCompiler] = set()
         for c in all_compilers_in_config(configuration):
```
Comment (johnwparent) [inline lib/spack/spack/solver/asp.py:2999]: @alalazo @scheibelp would this be the juncture where we would transition to using the new compiler interface? I.e. would we pull from the repo of installed spack packages and filter by compiler as well as composing package classes from the externals yaml? Or are we pulling compilers from the configs still? This is one of the several points where it's not clear how this new compiler interface is actually supposed to be integrated with the rest of Spack.

**Comment by johnwparent** [comment]
@alalazo Pushed the state pre modification for the compiler changes.
None of the changes in this PR relate to detection in any way, outside of the fact that it would be nice to be able to use them during detection, but none of the changes impact _how_ or _what_ is being detected. Basically, whenever the compiler packages are ready for real use in Spack (beyond detection) all the code here can just be copy and pasted to the new compiler objects with minimal refactors to make the class references correct.

File: lib/spack/spack/solver/asp.py [hunk 8]
```diff
@@ -316,8 +316,15 @@ def using_libc_compatibility() -> bool:
...
-    return compiler.compiler_verbose_output is not None
+def compiler_runs_for_lang(compiler) -> bool:
+    for lang in ["cc", "cxx", "fc", "f77"]:
+        comp_v_out_for_lang = compiler.try_compiler(lang)
+        if not comp_v_out_for_lang:
+            tty.debug(
+                f"the {lang} compiler {getattr(compiler, lang)} does not exist, or does not run correctly."
+                f" The compiler {compiler.spec} will not be used during concretization."
+            )
+        return comp_v_out_for_lang is not None
```
Comment (alalazo) [inline lib/spack/spack/solver/asp.py:328]: **TL;DR** The changes in this file are wrong. Can we revert them?

---

If the function was implemented correctly it would exclude compilers that define just `c` and `cxx` from concretization. Right now we only exclude a compiler object if there is no associated C compiler. I guess people still want to use `clang` without having a fortran compiler, so that would be a breaking behavior.

The function, though, is buggy as it iterates over 4 strings and exits at the first iteration. This has effectively the previous behavior on success: 
```python
>>> import spack.solver.asp
>>> import spack.compilers
>>> compilers = spack.compilers.all_compilers()
>>> compilers[0]
clang(/usr/bin/clang-12
     /usr/bin/clang++-12
     None
     None
     []
     ubuntu20.04)
>>> spack.solver.asp.compiler_runs_for_lang(compilers[0])
True
```
The failures, instead, are not treated gracefully:
```python
>>> compilers[0].cc = None
>>> spack.solver.asp.compiler_runs_for_lang(compilers[0])
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/solver/asp.py", line 321, in compiler_runs_for_lang
    comp_v_out_for_lang = compiler.try_compiler(lang)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/compiler.py", line 702, in try_compiler
    raise TryCompilerError(type(self).__name__, file, lang, err)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/compiler.py", line 797, in __init__
    with open(file, "r") as f:
         ^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: 'unknown'
```
This results in the following behavior if a single compiler does not provide the C language:
```console
$ spack spec zlib-ng
==> Error: [Errno 2] No such file or directory: 'unknown'
```
Overall, it doesn't seem much test went into this change :disappointed:

File: lib/spack/spack/compiler.py [hunk 9]
```diff
@@ -655,6 +677,55 @@ def search_regexps(cls, language):
...
+        return type(plat) in cls.supported_platforms
+
+    def try_compiler(self, lang, file=None, cmp_args=[], flags=[], definitions=[]):
+        """Try to use this compiler to compile a test file of specified language.
+        File is provided by caller or specified by language denotation.
+        Args:
+            compiler_path (str): path to compiler to be tested
+            file (str): path to file to be test compiled
+        Returns output of compilation if successful, Raises TryCompilerError otherwise
+        """
```
Comment (alalazo) [inline lib/spack/spack/compiler.py:690]: > Basically, whenever the compiler packages are ready for real use in Spack (beyond detection) all the code here can just be **copy and pasted** to the new compiler objects with minimal refactors to make the class references correct.

I still think it's better to encapsulate this logic, to avoid "cut and paste" later (and the accidental bugs that might come with it). As we agreed, I'll try to come up with a separate entity to try compile code and post it in some form. Will ping you in there.

File: lib/spack/spack/compiler.py [hunk 10]
```diff
@@ -454,36 +466,46 @@ def required_libs(self):
...
+    #         for flag_type in ["cflags" if cc == self.cc else "cxxflags", "cppflags", "ldflags"]:
+    #             cc_exe.add_default_arg(*self.flags.get(flag_type, []))
+
+    #         with self.compiler_environment():
+    #             return cc_exe(self.verbose_flag, fin, "-o", fout, output=str, error=str)
+    #     except spack.util.executable.ProcessError as pe:
+    #         tty.debug("ProcessError: Command exited with non-zero status: " + pe.long_message)
+    #         return None
+    #     finally:
+    #         shutil.rmtree(tmpdir, ignore_errors=True)
```
Comment (alalazo) [inline lib/spack/spack/compiler.py:507]:
```suggestion
```

File: lib/spack/spack/compiler.py (hunk 1 above)
Comment (alalazo) [reply lib/spack/spack/compiler.py:709]: These are good options on a separate class :wink:

File: lib/spack/spack/audit.py
Comment (alalazo) [inline lib/spack/spack/audit.py:1]: This wasn't tested either :disappointed: 
```console
$ spack config get compilers
compilers:
- compiler:
    spec: clang@=12.0.0
    paths:
      cc: null
      cxx: /usr/bin/clang++-12
      f77: null
      fc: null
    flags: {}
    operating_system: ubuntu20.04
    target: x86_64
    modules: []
    environment: {}
    extra_rpaths: []
- compiler:
    spec: gcc@=9.4.0
    paths:
      cc: /usr/bin/gcc
      cxx: /usr/bin/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    flags: {}
    operating_system: ubuntu20.04
    target: x86_64
    modules: []
    environment: {}
    extra_rpaths: []

$ spack audit configs
==> Error: 'spec'
$ spack -d audit configs
==> [2024-07-10-13:51:40.805408] Reading config from file /home/culpo/PycharmProjects/spack/etc/spack/defaults/config.yaml
==> [2024-07-10-13:51:40.825557] Imported audit from built-in commands
==> [2024-07-10-13:51:40.826958] Imported audit from built-in commands
==> [2024-07-10-13:51:40.827447] Reading config from file /home/culpo/.spack/linux/compilers.yaml
Traceback (most recent call last):
  File "/home/culpo/PycharmProjects/spack/bin/spack", line 52, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack_installable/main.py", line 42, in main
    sys.exit(spack.main.main(argv))
             ^^^^^^^^^^^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 1069, in main
    return _main(argv)
           ^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 1022, in _main
    return finish_parse_and_run(parser, cmd_name, args, env_format_error)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 1052, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 649, in _invoke_command
    return_val = command(parser, args)
                 ^^^^^^^^^^^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/cmd/audit.py", line 112, in audit
    subcommands[args.subcommand](parser, args)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/cmd/audit.py", line 59, in configs
    reports = spack.audit.run_group(args.subcommand)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/audit.py", line 153, in run_group
    errors = run_check(check, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/audit.py", line 168, in run_check
    return CALLBACKS[tag].run(**kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/audit.py", line 136, in run
    errors.extend(fn(**kwargs))
                  ^^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/audit.py", line 217, in _search_broken_compilers
    spec = spack.spec.parse_with_version_concrete(compiler["spec"])
                                                  ~~~~~~~~^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/external/_vendoring/ruamel/yaml/comments.py", line 927, in __getitem__
    return ordereddict.__getitem__(self, key)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyError: 'spec'
```

File: lib/spack/spack/compiler.py [hunk 11]
```diff
@@ -707,6 +778,28 @@ def compiler_environment(self):
             os.environ.update(backup_env)
 
 
+class CompilerLang:
+    CC = "cc"
+    CXX = "cxx"
+    FC = "fc"
+    F77 = "f77"
+    ext_for_lang = {CC: "c", CXX: "cxx", FC: "f", F77: "f77"}
```
Comment (alalazo) [inline lib/spack/spack/compiler.py:785]: ?

File: lib/spack/spack/compilers/__init__.py [hunk 12]
```diff
@@ -902,10 +902,11 @@ def _default(fn_args):
...
         except spack.util.executable.ProcessTimeoutError as e:
             error = f"Couldn't get version for compiler {path}\n" + str(e)
+        except spack.util.executable.ProcessError as e:
+            error = f"Couldn't get version for compiler {path}\n" + str(e)
```
Comment (alalazo) [inline lib/spack/spack/compilers/__init__.py:909]: (minor) I don't get this. One is a subclass of the other. The error message produced seems to be exactly the same. Why are we not deleting the subclass?

**Review by alalazo — CHANGES_REQUESTED** [review]
I didn't check all the details. Honestly I'm a bit perplexed by the state of the PR. I thought this was on the blocking path for Windows, but:
1. Nothing works (`spack audit`, the check in `asp.py`, see below for details)
2. The features that are advertized seem to be generic, and not Windows specific

As promised, I'll try to experiment with some API to try compile code that can be reused both in `Compiler` and in `Packages`.

EDIT: see https://github.com/spack/spack/pull/43458#discussion_r1672412803

File: lib/spack/spack/compilers/__init__.py (hunk 12 above)
Comment (johnwparent) [reply lib/spack/spack/compilers/__init__.py:909]: Unsure, I didn't want to deal with it here so I just fixed what was already there. IMO there's value in knowing when a compiler times out vs actually fails to execute, so I can update the `ProcessTimeoutError` context error message to report said failure as opposed to just avoiding catching the subclass explicitly.
Thoughts?

File: lib/spack/spack/audit.py
Comment (johnwparent) [reply lib/spack/spack/audit.py:1]: Ok this one was 100% working before I refactored and then un-refactored based on your earlier comments and our discussion yesterday, some fix must have been dropped in the churn because this is _how_ I've been testing this so far.
Probably a very simple fix.

File: lib/spack/spack/compiler.py (hunk 11 above)
Comment (johnwparent) [reply lib/spack/spack/compiler.py:785]: Cruft from an earlier pass. This PR wasn't actually ready to go in when you halted progress on it, and I don't like to remove code until I'm 100% sure I'm not going to need it again.

File: lib/spack/spack/compiler.py (hunk 10 above)
Comment (johnwparent) [reply lib/spack/spack/compiler.py:507]: Yes, dead code as I wanted to refer back to this in case there was an issue with the `asp.py` changes. Like I mention in another comment, this PR wasn't ready to go in, I was in the middle of an iteration when work was paused, so I left a lot of cruft in in case I needed to refer to this.

File: lib/spack/spack/solver/asp.py (hunk 8 above)
Comment (johnwparent) [reply lib/spack/spack/solver/asp.py:328]: This was not actually finished when you left your comment that paused my work on this, so admittedly, testing was exceptionally minimal, and only on compilers with all four compilers defined (i.e. my local MSVC configuration). 

Why would we revert instead of just _fixing_ the issues? 
I'll fix them, and then iterate from there.

File: lib/spack/spack/compiler.py (hunk 9 above)
Comment (johnwparent) [reply lib/spack/spack/compiler.py:690]: Would we not just be cut and pasting from the shared interface eventually anyway? 

> As we agreed, I'll try to come up with a separate entity to try compile code and post it in some form

I'll take a pass at the try compiler object today, this PR has stagnated for a long time, and I'd like to continue making forward progress and it doesn't really make sense for both of us to work on the same thing.

**Comment by johnwparent** [comment]
@alalazo 
> I didn't check all the details. Honestly I'm a bit perplexed by the state of the PR. I thought this was on the blocking path for Windows, but:

Not sure what you mean by blocking path for Windows...

> 1. Nothing works (`spack audit`, the check in `asp.py`, see below for details)

This is mostly due to a month of stagnation and a series of messy rebases, the code was not ready when work was interrupted, and was just refactored enough so it was readable. The errors are _very_ simple to resolve.

> 2. The features that are advertized seem to be generic, and not Windows specific

Nowhere in this PR, or in our discussion is it mentioned that this is for Windows specifically, in fact the motivating use case is Clang on Ubuntu.

> As promised, I'll try to experiment with some API to try compile code that can be reused both in `Compiler` and in `Packages`.

I'll just do this as part of iterating on this PR as there's really nothing else for me to do to make forward progress as there's not a lot else here that I can iterate on, also I think it makes the first sense for me to take the first pass at another class as I've been working with this concept for a little while now.

File: lib/spack/spack/compiler.py (hunk 11 above)
Comment (alalazo) [reply lib/spack/spack/compiler.py:785]: @johnwparent Didn't we agree yesterday that I would review this PR asap, as soon as you ping me in here, when you thought that was ready for review? I saw the ping, and I reviewed the PR :thinking: Now you say it's not ready... fine. Ping me when it's ready and I'll re-review.

**Comment by alalazo** [comment]
> I'll just do this as part of iterating on this PR as there's really nothing else for me to do to make forward progress as there's not a lot else here that I can iterate on.

Okay, thanks. I thought we also agreed yesterday on that, but probably there was some misunderstanding.

**Comment by johnwparent** [comment]
> > I'll just do this as part of iterating on this PR as there's really nothing else for me to do to make forward progress as there's not a lot else here that I can iterate on.
> 
> Okay, thanks. I thought we also agreed yesterday on that, but probably there was some misunderstanding.

I had thought we left it at: "you will review what is there and then we decide on an action plan based on your review" so, yes it seems we had different ideas from that meeting 🤷

File: lib/spack/spack/compiler.py (hunk 11 above)
Comment (johnwparent) [reply lib/spack/spack/compiler.py:785]: @alalazo 
I think we clearly just misunderstood each other. When dev paused here, the changes were still a WIP, and undergoing active review. When we discussed this PR yesterday, and we agreed that we'd get this PR ready ASAP, I understood that to mean I would get it into the rough state it was in prior to your comment and my overhaul to the compiler package approach (with any modification required to update detection) and we'd move from there, not that I would get into a "ready to merge right now" state.
Apologies if that wasn't what you had intended, we should have clarified what this review would be for, I did not mean to  mislead you as to the state of this PR, again, my apologies.
, I think we just had different expectations for what this would look like when you reviewed it, I meant for you to look at the approach from a conceptual perspective to see about the approach moving forward and provide high level comments as you have a better picture of where the rest of the compiler work will be going in the future, not to review the implementation details.

Comment (alalazo) [reply lib/spack/spack/compiler.py:785]: > Apologies if that wasn't what you had intended, we should have clarified what this review would be for, I did not mean to mislead you as to the state of this PR, again, my apologies.

No worries, misunderstanding happens (often, apparently, in this PR :grimacing: ). I'm happy to re-review when ready.

Comment (johnwparent) [reply lib/spack/spack/compiler.py:785]: > misunderstanding happens (often, apparently, in this PR 😬 )

I think we're on the same page now! (I hope 👍 )

File: lib/spack/spack/compilers/msvc.py
Comment (alalazo) [inline lib/spack/spack/compilers/msvc.py:1]: @johnwparent The code in here prepares an environment to use `msvc`. Can we move it to `setup_run_environment` in the `Msvc` package?

**Comment by alalazo** [comment]
Superseded by #45189 Tests for compilers are now "normal" tests on the package. Can we close this PR?

**Comment by johnwparent** [comment]
> Superseded by #45189 Tests for compilers are now "normal" tests on the package. Can we close this PR?

@alalazo do we actually test to see if any of the compilers can compile a stub "hello world" before writing them to the packages yaml/using them? That was the initial motivation behind this PR, and I don't see any compilers implementing tests/attempts to use them

I'm fine with closing this PR since the interface has changed so much, and I already have a follow-up PR written.

**Comment by alalazo** [comment]
> do we actually test to see if any of the compilers can compile a stub "hello world" before writing them to the packages yaml/using them?

We try to compile an "hello world", at least on linux, to check which libc is supported by a candidate compiler. The check is currently done on compilers that are available "locally" (either externals, or installed in the local store). For buildcaches we just infer a libc compatibility, under the assumption we don't deal with cross-compilers yet.
